File: python/examples/native/edit/edit_factory.py
#!/usr/bin/env python3
"""
Edit Tool - Factory Pattern with Interactive Approval

Example: Edit and modify content using Matimo's edit tool
Demonstrates editing and modifying file contents with interactive approval
"""

import asyncio
import sys
import tempfile
from pathlib import Path

from dotenv import load_dotenv

from matimo import Matimo, get_global_approval_handler


def create_approval_callback():
    """Create an interactive approval callback for file operations"""

    async def approval_handler(request):
        """
        Handle approval requests interactively
        
        Args:
            request: ApprovalRequest with tool_name, description, params
            
        Returns:
            bool: True if approved, False otherwise
        """
        is_interactive = sys.stdin.isatty()

        print('\n' + '=' * 70)
        print('🔒 APPROVAL REQUIRED FOR FILE OPERATION')
        print('=' * 70)
        print(f'\n📋 Tool: {request.tool_name}')
        print(f'📝 Description: {request.description or "(no description provided)"}')
        print('\n📝 File Operation:')
        print(f'   Path: {request.params.get("filePath", "N/A")}')
        print(f'   Operation: {request.params.get("operation", "N/A")}')
        
        if request.params.get('startLine'):
            print(f'   Start Line: {request.params["startLine"]}')
        if request.params.get('endLine'):
            print(f'   End Line: {request.params["endLine"]}')

        if not is_interactive:
            print('\n❌ REJECTED - Non-interactive environment (no terminal)')
            print('\n💡 To enable auto-approval in CI/scripts:')
            print('   export MATIMO_AUTO_APPROVE=true')
            print('\n💡 Or approve specific patterns:')
            print('   export MATIMO_APPROVED_PATTERNS="edit"')
            print('\n' + '=' * 70 + '\n')
            return False

        # Interactive mode: prompt user
        print('\n❓ User Action Required')
        user_input = input('   Type "yes" to approve or "no" to reject: ').strip().lower()
        approved = user_input in ('yes', 'y')

        if approved:
            print('   ✅ Operation APPROVED by user')
        else:
            print('   ❌ Operation REJECTED by user')
        print('=' * 70 + '\n')

        return approved

    return approval_handler


async def main():
    """
    Main entry point for edit factory example
    
    Demonstrates:
    1️⃣ Replacing content in a file (replace line 2)
    2️⃣ Inserting new content into a file
    """
    # Load environment variables
    load_dotenv()

    # Initialize Matimo with auto-discovery to find all tools
    matimo = await Matimo.init(auto_discover=True)

    # Configure centralized approval handler
    approval_handler = get_global_approval_handler()
    approval_handler.set_approval_callback(create_approval_callback())

    print('=== Edit Tool - Factory Pattern (Interactive Approval) ===\n')

    # Create a temporary file for demonstration
    temp_file = None
    try:
        # Create temp file with initial content
        with tempfile.NamedTemporaryFile(
            mode='w',
            suffix='.txt',
            delete=False,
            dir='/tmp'
        ) as f:
            temp_file = Path(f.name)
            f.write('Line 1\nLine 2\nLine 3\n')

        # 1️⃣ Example 1: Replace text in file (replace line 2)
        print('1️⃣ Replacing content in file\n')
        print('Original content:')
        with open(temp_file, 'r') as f:
            original_content = f.read()
        print(original_content)
        print('---\n')

        try:
            result = await matimo.execute(
                'edit',
                {
                    'filePath': str(temp_file),
                    'operation': 'replace',
                    'content': 'Line 2 (Modified)',
                    'startLine': 2,
                    'endLine': 2,
                }
            )

            if result.get('success'):
                print(f'Edit Result: {result.get("success")}')
                print(f'Lines Affected: {result.get("linesAffected")}')
                print('\nModified content:')
                with open(temp_file, 'r') as f:
                    modified_content = f.read()
                print(modified_content)
            else:
                print(f'Edit denied: {result.get("error")}')
        except Exception as e:
            print(f'Error: {e}')
        
        print('---\n')

        # 2️⃣ Example 2: Insert new content
        print('2️⃣ Inserting new line\n')
        
        try:
            insert_result = await matimo.execute(
                'edit',
                {
                    'filePath': str(temp_file),
                    'operation': 'insert',
                    'content': 'New inserted line',
                    'startLine': 2,
                }
            )

            if insert_result.get('success'):
                print(f'Insert Result: {insert_result.get("success")}')
                print(f'Lines Affected: {insert_result.get("linesAffected")}')
                print('\nFinal content:')
                with open(temp_file, 'r') as f:
                    final_content = f.read()
                print(final_content)
            else:
                print(f'Insert denied: {insert_result.get("error")}')
        except Exception as e:
            print(f'Error: {e}')
        
        print('---\n')

    except Exception as error:
        print(f'Error editing file: {error}')
        if hasattr(error, 'message'):
            print(f'Error message: {error.message}')
    
    finally:
        # Clean up temporary file
        if temp_file and temp_file.exists():
            try:
                temp_file.unlink()
            except Exception as e:
                print(f'Warning: Could not delete temp file: {e}')


if __name__ == '__main__':
    asyncio.run(main())

File: typescript/examples/tools/edit/edit-factory.ts
import { MatimoInstance, getGlobalApprovalHandler, type ApprovalRequest } from '@matimo/core';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import * as readline from 'readline';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

/**
 * Create an interactive approval callback for file operations
 */
function createApprovalCallback() {
  return async (request: ApprovalRequest): Promise<boolean> => {
    const isInteractive = process.stdin.isTTY;

    console.info('\n' + '='.repeat(70));
    console.info('🔒 APPROVAL REQUIRED FOR FILE OPERATION');
    console.info('='.repeat(70));
    console.info(`\n📋 Tool: ${request.toolName}`);
    console.info(`📝 Description: ${request.description || '(no description provided)'}`);
    console.info(`\n📄 File Operation:`);
    console.info(`   Path: ${request.params.filePath}`);
    if (request.params.startLine) {
      console.info(`   Start Line: ${request.params.startLine}`);
    }
    if (request.params.endLine) {
      console.info(`   End Line: ${request.params.endLine}`);
    }

    if (!isInteractive) {
      console.info('\n❌ REJECTED - Non-interactive environment (no terminal)');
      console.info('\n💡 To enable auto-approval in CI/scripts:');
      console.info('   export MATIMO_AUTO_APPROVE=true');
      console.info('\n💡 Or approve specific patterns:');
      console.info('   export MATIMO_APPROVED_PATTERNS="edit"');
      console.info('\n' + '='.repeat(70) + '\n');
      return false;
    }

    // Interactive mode: prompt user
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
    });

    return new Promise((resolve) => {
      console.info('\n❓ User Action Required');
      const question = '   Type "yes" to approve or "no" to reject: ';

      rl.question(question, (answer) => {
        const approved = answer.toLowerCase() === 'yes' || answer.toLowerCase() === 'y';

        if (approved) {
          console.info('   ✅ Operation APPROVED by user');
        } else {
          console.info('   ❌ Operation REJECTED by user');
        }
        console.info('='.repeat(70) + '\n');

        rl.close();
        resolve(approved);
      });
    });
  };
}

/**
 * Example: Edit tool using factory pattern
 * Demonstrates editing and modifying file contents with interactive approval
 */
async function editExample() {
  // Initialize Matimo with autoDiscover to find all tools (core + providers)
  const matimo = await MatimoInstance.init({ autoDiscover: true });

  // Configure centralized approval handler
  const approvalHandler = getGlobalApprovalHandler();
  approvalHandler.setApprovalCallback(createApprovalCallback());

  console.info('=== Edit Tool - Factory Pattern (Interactive Approval) ===\n');

  // Create a temp file for demonstration
  const tempFile = path.join(__dirname, 'temp-demo.txt');
  fs.writeFileSync(tempFile, 'Line 1\nLine 2\nLine 3\n');

  try {
    // Example 1: Replace text in file (replace line 2)
    console.info('1. Replacing content in file\n');
    console.info('Original content:');
    console.info(fs.readFileSync(tempFile, 'utf-8'));
    console.info('---\n');

    const result = await matimo.execute('edit', {
      filePath: tempFile,
      operation: 'replace',
      content: 'Line 2 (Modified)',
      startLine: 2,
      endLine: 2,
    });

    if ((result as any).success) {
      console.info('Edit Result:', (result as any).success);
      console.info('Lines Affected:', (result as any).linesAffected);
      console.info('\nModified content:');
      console.info(fs.readFileSync(tempFile, 'utf-8'));
    } else {
      console.info('Edit denied:', (result as any).error);
    }
    console.info('---\n');

    // Example 2: Insert new content
    console.info('2. Inserting new line\n');
    const insertResult = await matimo.execute('edit', {
      filePath: tempFile,
      operation: 'insert',
      content: 'New inserted line',
      startLine: 2,
    });

    if ((insertResult as any).success) {
      console.info('Insert Result:', (insertResult as any).success);
      console.info('Lines Affected:', (insertResult as any).linesAffected);
      console.info('\nFinal content:');
      console.info(fs.readFileSync(tempFile, 'utf-8'));
    } else {
      console.info('Insert denied:', (insertResult as any).error);
    }
    console.info('---\n');
  } catch (error: any) {
    console.error('Error editing file:', error.message);
  } finally {
    // Clean up temp file
    if (fs.existsSync(tempFile)) {
      fs.unlinkSync(tempFile);
    }
  }
}

editExample();
